mis_funciones.cpp

```diff
@@ -0,0 +1,158 @@
+/**
+* @file mis_funciones.cpp
+* @version 1.0
+* @date 18/7/24
+* @author Ramos Leonardo
+*/
+#include "mis_funciones.hpp"
+
+/// @cond VARIABLES EXTERNAS
+extern BusOut keypadRows;                         
+extern BusIn keypadCols;                         
+extern DigitalOut workingLed;                     
+extern DigitalOut accessGrantedLed;              
+extern DigitalOut accessDeniedLed;               
+extern UnbufferedSerial uartUSB;                 
+
+extern bool sequenceDetectedState;                      
+extern bool accessState;                                
+extern bool timeoutState;                               
+extern int colReading;                                
+extern int rowReading;                                  
+extern uint8_t currentIndex;                            
+extern uint16_t accumulatedTime;                        
+extern char inputSequence[SEQUENCE_LEN + 1];            
+extern char keypadAllowedEntries[ROW_LEN][COL_LEN];    
+/// @endcond
+
+/**
+ * @brief keypadInit Inicializa el keypad 4x4 para su uso. Principalmente
+ *        configura los GPIO de columnas como input PULL DOWN y da valores iniciales
+ *        a ciertos flags. 
+ */
+void keypadInit() {
+  keypadCols.mode(PullDown);
+  inputSequence[SEQUENCE_LEN] = '\0';
+  sequenceDetectedState = false;
+  currentIndex = 0;
+  accumulatedTime = 0;
+}
+
+/**
+ * @brief uartInit Configura velocidad de transmisión y formato de la trama
+ *        de datos de UART
+ */
+void uartInit() {
+  uartUSB.baud(9600);
+  uartUSB.format(/*bits*/ 8,
+                 /*paridad*/ SerialBase::None,
+                 /*bits de stop*/ 1);
+}
+
+/**
+ * @brief keypadSweepUpdate Barre el keypad 4x4 hasta encontrar
+ *        una tecla presionada, guarda el caracter asociado
+ *        a esa tecla e indica mediante flag sequenceDetectedState
+ *        que se ha ingresado la secuencia completa
+ */
+void keypadSweepUpdate() {
+  for (uint8_t i = 0; i < ROW_LEN; i++) {
+    keypadRows = (1 << i);
+    if (keypadCols != 0) {
+      delay(100);
```

Comment:
Otro bloqueo infernal.
Esta función demora 400ms en ejecutarse, 100 durmiendo en cada fila

Reply:
Me puedo equivocar porque no sé nada de Mbed. Entiendo que keypadCols es un objeto, porque líneas abajo invocás el método `read()` para obtener el valor. O sea, que esto SIEMPRE es != 0 porque el objecto existe, o sea que tu intento de hacer algo parecido a un anti-rebote no funciona.

Reply:
El objeto guarda el estado de cada GPIO (bit 0, gpio0, bit1, gpi01, etc), por lo cual si todos los gpio están en cero, efectivamente el valor es cero. Si es cierto que podría ser más específico el antirebote, pues no se especifica la columna. Los 100 ms de delay lo tuve que ajustar por el rebote del keypad , y se ejecuta solo de detectar un dígito, no en cada fila (luego se rompe el for). Otra forma de resolverlo es con una máquina de estado, pero ese tema viene luego de este tp.

Reply:
Respecto a mi comentario, es incorrecto.
https://os.mbed.com/docs/mbed-os/v6.16/apis/busin.html
La clase redefine un operador sólo para molestar a los que conocemos al resto del mundo. Usar el objeto en un contexto de enteros invoca el método read().

Yo lo que hago en circunstancias como ésta (el anti-rebote, no las locuras de Mbed...) es hacer 2 lecturas no bloqueantes. Tengo una cuenta de tiempo, hago la primera lectura, veo que hay algo, sigo con mi vida y marco que x ms después hay que verificar que se haya leído lo mismo. Eso "técnicamente" es una máquina de estados, pero con sólo setear un flag se resuelve.
Otro detalle: si se suelta la tecla en el intervalo entre que validaste que no hay rebote y el que leés de vuelta, te queda una inconsistencia de la que tenés que salir como si no hubiera pasado nada. O sea, se puede dar la condición de que veas, confirmes, y cuando scaneás ya no está. Eso se resuelve escaneando 2 veces en vez de mirar el bus, que es lento, o admitiendo que puede no haber nada y siguiendo de largo. PERO, si metés ese bloqueo, es caro en tiempo de ejecución.